fix(minifier): should not merge conditional function calls if referencing the function has a side-effect

sapphi-red opened the pull request:
`a ? b(c, d) : b(e, d)` should not be compressed to `b(a ? c : e, d)` if `b` has a side effect.
Because `b` may update `a` variable.

For example, compressing the following code changes the behavior.
```js
var foo = () => console.log('foo')
var bar = () => { foo = () => console.log('foo2') }
bar() ? foo(1, 2) : foo(3, 2) // outputs 'foo'

// was compressed into
var foo = () => console.log('foo')
var bar = () => { foo = () => console.log('foo2') }
foo(bar() ? 1 : 3, 2); // outputs 'foo2'
```

We can improve this if we know that `foo` is not updated after it was declared (i.e. `foo` is a const). But I didn't implement it for now.

<details>
<summary>evaluation steps in detail</summary>

Before the compression, the code is evaluated in the following steps:
1. Evaluate `a`
1. Evaluate `b`
1. Evaluate `c` / `e`
1. Evaluate `d`
1. Evaluate `b(c,d)` / `b(e,d)`

After the compression, the code is evaluated in the following steps:
1. Evaluate `b`
1. Evaluate `a`
1. Evaluate `c` / `e`
1. Evaluate `d`
1. Evaluate `b(c,d)` / `b(e,d)`

The steps of 1 and 2 are reversed.

</details>

**References**
- [Spec of conditional expression](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#sec-conditional-operator-runtime-semantics-evaluation)
- [Spec of call expression](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#sec-function-calls-runtime-semantics-evaluation)

sapphi-red commented:
* **#8922** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8922?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8922?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8922).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 commented on crates/oxc_minifier/tests/peephole/mod.rs:
```diff
@@ -22,7 +22,7 @@ fn integration() {
     })",
         r#"require("./index.js")(function(e, os) {
-    return console.log(e || JSON.stringify(os));
+    return e ? console.log(e) : console.log(JSON.stringify(os));
```
how come this changed? `e` isn't a global so shouldn't have a side effect right?

sapphi-red replied:
It's not the case here, but `console.log` may change `e`.
```js
var console = { get log() { e = false; return () => {} } }
return e ? console.log(e) : console.log(JSON.stringify(os));
// this cannot be compressed to `return console.log(e ? e : JSON.stringify(os));`
```

camc314 replied:
ohhh gotcha, thanks for explaining.

Boshen commented:
### Merge activity

* **Feb 6, 10:06 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

sapphi-red committed:
fix(minifier): should not merge conditional function calls if referencing the function has a side-effect (#8922)

`a ? b(c, d) : b(e, d)` should not be compressed to `b(a ? c : e, d)` if `b` has a side effect.
Because `b` may update `a` variable.

For example, compressing the following code changes the behavior.
```js
var foo = () => console.log('foo')
var bar = () => { foo = () => console.log('foo2') }
bar() ? foo(1, 2) : foo(3, 2) // outputs 'foo'

// was compressed into
var foo = () => console.log('foo')
var bar = () => { foo = () => console.log('foo2') }
foo(bar() ? 1 : 3, 2); // outputs 'foo2'
```

We can improve this if we know that `foo` is not updated after it was declared (i.e. `foo` is a const). But I didn't implement it for now.

<details>
<summary>evaluation steps in detail</summary>

Before the compression, the code is evaluated in the following steps:
1. Evaluate `a`
1. Evaluate `b`
1. Evaluate `c` / `e`
1. Evaluate `d`
1. Evaluate `b(c,d)` / `b(e,d)`

After the compression, the code is evaluated in the following steps:
1. Evaluate `b`
1. Evaluate `a`
1. Evaluate `c` / `e`
1. Evaluate `d`
1. Evaluate `b(c,d)` / `b(e,d)`

The steps of 1 and 2 are reversed.

</details>

**References**
- [Spec of conditional expression](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#sec-conditional-operator-runtime-semantics-evaluation)
- [Spec of call expression](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#sec-function-calls-runtime-semantics-evaluation)
